src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java

```diff
@@ -254,4 +281,98 @@ private void handleIfNecessary() {
       }
     }
   }
+
+  /**
+   * Decorator used to avoid selectors from being blocked reading from not yet written streams.
+   *
+   */
+  private class DecoratedPipedInputStream extends PipedInputStream {
+
+    private AtomicBoolean wasDataWritten = new AtomicBoolean(false);
+    private AtomicBoolean isWriterClosed = new AtomicBoolean(false);
+
+    DecoratedPipedInputStream(PipedOutputStream output, int bufferSize) throws IOException {
+      super(output, bufferSize);
+    }
+
+    /**
+     * Same as parent except that it returns <code>0</code> if nobody wrote in the other side of the stream yet and the
+     * stream is not closed.
+     *
+     * @param b Buffer to fill by the method.
+     * @return The number of bytes read if available;
+     *         <code>-1</code> if the stream is closed;
+     *         <code>0</code> if nobody wrote data in the other side of the stream.
+     */
+    @Override
+    public int read(byte[] b) throws IOException {
+      if (!wasDataWritten.get() && !isWriterClosed.get()) {
+        return 0;
+      }
+      return super.read(b);
+    }
+
+    /**
+     * Same as parent except that it returns <code>0</code> if nobody wrote in the other side of the stream yet and the
+     * stream is not closed.
+     *
+     * @param b Buffer to fill by the method.
+     * @return The number of bytes read if available;
+     *         <code>-1</code> if the stream is closed;
+     *         <code>0</code> if nobody wrote data in the other side of the stream.
+     */
+    @Override
+    public synchronized int read(byte[] b, int off, int len) throws IOException {
+      if (!wasDataWritten.get() && !isWriterClosed.get()) {
+        return 0;
+      }
+      return super.read(b, off, len);
+    }
+
+    void setWriterClosed(boolean value) {
+      this.isWriterClosed.set(value);
+    }
+
+    void setDataWasWritten(boolean value) {
+      wasDataWritten.set(value);
+    }
+  }
+
+  private class DecoratedPipedOutputStream extends PipedOutputStream {
+
+    private DecoratedPipedInputStream countingSink;
+
+    @Override
+    public synchronized void connect(PipedInputStream snk) throws IOException {
+      super.connect(snk);
+      if (!(snk instanceof DecoratedPipedInputStream)) {
+        throw new IllegalArgumentException("Sink must be an instance of CountingPipedInputStream");
```

Comment:
shouldn't this be DecoratedPipedInputStream?

Reply:
Obviously!